### ISSUE TITLE
Model with softmax output layer not outputing probabilities but just 1 and 0 

### ISSUE THREAD

**Issue by croraf** [tensorflow/tfjs#8271]
I'm testing the model provided in node-js example https://github.com/tensorflow/tfjs-examples/blob/master/mnist-node/model.js to classify images of numbers into classes 0 to 9.

It is a `tf.sequential()` model, and its final layer is `model.add(tf.layers.dense({ units: 10, activation: "softmax" }));`

Therefore the prediction should output probabilities per class, but this is not happening. Instead only 0s and 1s (in a one-hot fashion) are outputed  in the following logging.

  ```ts
const predictions = model.predict(data);
predictions.print();
```
![image](https://github.com/tensorflow/tfjs/assets/19698636/47e1decb-3e98-46f4-a7ef-2bef1729d077)

How to get the probabilities of specific image belonging to specific class

**Comment by gaikwadrahul8** [comment]
Hi, @croraf 

I apologize for the delayed response and you're absolutely right. The final layer in your model `(tf.layers.dense({ units: 10, activation: 'softmax' }))` is configured correctly to output probabilities for each class (0-9). However, the way you're interpreting the model's prediction `(predictions.print())` might be misleading. maybe you're seeing only 0s and 1s because
One-Hot Encoded Predictions format, the softmax activation ensures the output is a probability distribution between 0 and 1, where each element represents the probability of the input belonging to a specific class.When you print the tensor using `predictions.print()`, TensorFlow might display the values in a one-hot encoded format. This means the element with the highest probability will be set to 1 and all others will be 0.

Could you please try the below code snippet and see if it is working as expected or not ? If the issue still persists please let us know.  Thank you for your cooperation and patience. 

```
const predictions = model.predict(data);
const classProbabilities = predictions.dataSync(); // Use dataSync() for Node.js

// Loop through each element (probability for each class)
for (let i = 0; i < classProbabilities.length; i++) {
  console.log(`Probability of class ${i}: ${classProbabilities[i]}`);
}

```

**Comment by croraf** [comment]
It outputs the same (except in one value where it is basically 0)
![image](https://github.com/tensorflow/tfjs/assets/19698636/37c3334d-4b75-4677-a871-10968ff36fb2)

**Comment by gaikwadrahul8** [comment]
Hi, @croraf 

I apologize for the delayed response, could you please try something like below and see is it working as expected or not? if now please help me with your Github repo to replicate the same behavior from our end to investigate this issue further.

```
const predictions = model.predict(data);
const classProbabilities = predictions.dataSync(); // Use dataSync() for Node.js

// Loop through each element (probability for each class)
for (let i = 0; i < classProbabilities.length; i++) {
console.log(`Probability of class ${i}: ${classProbabilities[i].toFixed(4)}`);
}
```

Thank you for your cooperation and patience.

**Comment by croraf** [comment]
Hi @gaikwadrahul8 . It is the same. I get:
![image](https://github.com/tensorflow/tfjs/assets/19698636/cb0f66ba-4071-4e1d-9ccb-607b1fe9f992)

Here is my repo: https://github.com/croraf/x-ray-ai (it has a README)

After training `npm run train``, you can use `npm run predict` to reproduce the issue.

The repo is basically a reproduction of example in `tfjs` https://github.com/tensorflow/tfjs-examples/blob/master/mnist-node/model.js

**Comment by Varun-S10** [comment]
Hi @croraf, apologies for the delay. Thank you for sharing the code to reproduce the issue. I have tried running your code for training and it returns (0 , util_1.isNullOrUndefined) is not a function. I have also attempted to run the tfjs-examples/mnist-node and encountered the same problem. I will escalate this to the team for further investigation. Thank you for your patience.

**Comment by buffcode** [comment]
@Varun-S10 thats because of https://github.com/tensorflow/tfjs/issues/8609, the patch is merged but not released, so its still not compatible with recent node versions

**Comment by Varun-S10** [comment]
Hi @croraf, Thank you for bringing this to our attention and for providing a comprehensive overview. We appreciate the time you invested in creating this issue.

Currently, our internal development team is concentrated on other scheduled milestones, so we cannot prioritize this task immediately. However, we would love to see this move forward as a community-led effort.

If you’d like to submit a fix, please fork the repo and open a PR following the [TensorFlow Contribution Guidelines](https://www.tensorflow.org/community/contribute), [Tfjs Contributions ](https://github.com/tensorflow/tfjs/blob/master/tfjs-vis/CONTRIBUTING.md#how-to-contribute)and be sure to link this issue in your PR to avoid duplicate work, kindly leave a comment here if you plan to take this on.

Thank you again for your interest and time.
